<h1> ACTIVITY1 - PALMER PENGUINS DATA ANALYTICS PROJECT </h1>

<h2> Analyst : John Lennon Jarabese</h2>

In [2]:
import numpy as np

In [3]:
import pandas as pd

In [4]:
import warnings
warnings.filterwarnings('ignore')

In [5]:
penguins = pd.read_csv('datasets\\penguins.csv')

In [6]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

<h1>Descriptive Analytics</h1>

In [7]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

<h1>Filtering a Column / Data</h1>

In [8]:
penguins['species']

0         Adelie
1         Adelie
2         Adelie
3         Adelie
4         Adelie
         ...    
339    Chinstrap
340    Chinstrap
341    Chinstrap
342    Chinstrap
343    Chinstrap
Name: species, Length: 344, dtype: object

In [9]:
penguins['sex']

0        male
1      female
2      female
3         NaN
4      female
        ...  
339      male
340    female
341      male
342      male
343    female
Name: sex, Length: 344, dtype: object

In [10]:
penguins[penguins['species'] == "Adelie"]

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
149  Adelie      Dream            37.8           18.1              193.0   
150  Adelie      Dream            36.0           17.1              187.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
147       3475.0  female  2009  
148       3450.0  female  2009  
149       3750.0    male  2009  
150       3700.0  female  2009  
151       4000.0    male  2009  

[152 rows x 8 columns]

Another Example

In [11]:
adelie_df = penguins[penguins['species'] == "Adelie"]

In [12]:
gentoo_df = penguins[penguins['species'] == "Gentoo"]

In [13]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

In [14]:
adelie_df

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
147  Adelie      Dream            36.6           18.4              184.0   
148  Adelie      Dream            36.0           17.8              195.0   
149  Adelie      Dream            37.8           18.1              193.0   
150  Adelie      Dream            36.0           17.1              187.0   
151  Adelie      Dream            41.5           18.5              201.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
147       3475.0  female  2009  
148       3450.0  female  2009  
149       3750.0    male  2009  
150       3700.0  female  2009  
151       4000.0    male  2009  

[152 rows x 8 columns]

In [15]:
gentoo_df

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
152  Gentoo  Biscoe            46.1           13.2              211.0   
153  Gentoo  Biscoe            50.0           16.3              230.0   
154  Gentoo  Biscoe            48.7           14.1              210.0   
155  Gentoo  Biscoe            50.0           15.2              218.0   
156  Gentoo  Biscoe            47.6           14.5              215.0   
..      ...     ...             ...            ...                ...   
271  Gentoo  Biscoe             NaN            NaN                NaN   
272  Gentoo  Biscoe            46.8           14.3              215.0   
273  Gentoo  Biscoe            50.4           15.7              222.0   
274  Gentoo  Biscoe            45.2           14.8              212.0   
275  Gentoo  Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  year  
152       4500.0  female  2007  
153       5700.0    male  2007  
154       4450.0  female  2007  
155       5700.0    male  2007  
156       5400.0    male  2007  
..           ...     ...   ...  
271          NaN     NaN  2009  
272       4850.0  female  2009  
273       5750.0    male  2009  
274       5200.0  female  2009  
275       5400.0    male  2009  

[124 rows x 8 columns]

In [16]:
adelie_df['body_mass_g'].mean()

np.float64(3700.662251655629)

In [17]:
adelie_df['bill_length_mm'] == 30.1

0      False
1      False
2      False
3      False
4      False
       ...  
147    False
148    False
149    False
150    False
151    False
Name: bill_length_mm, Length: 152, dtype: bool

In [18]:
(penguins['body_mass_g'] <= 4000 ) & (penguins['sex'] == 'male')

0       True
1      False
2      False
3      False
4      False
       ...  
339     True
340    False
341     True
342    False
343    False
Length: 344, dtype: bool

<h1> Cleaning the data Frame</h1>

In [19]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h1>What to do with all the NULL VALUES?</h1>

In [20]:
penguins['bill_length_mm'].mean()

np.float64(43.9219298245614)

In [21]:
penguins['sex'].mode()

0    male
Name: sex, dtype: object

<h1>How to FILLUP NULL VALUES?</h1>

In [22]:
penguins['bill_length_mm'] = penguins['bill_length_mm'].fillna(penguins['bill_length_mm'].mean(0))

In [23]:
penguins['sex'] = penguins['sex'].fillna(penguins['sex'].mode()[0])

In [24]:
penguins.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        2
flipper_length_mm    2
body_mass_g          2
sex                  0
year                 0
dtype: int64

<h1>DATA TYPES</h1>

In [25]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     344 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                344 non-null    object 
 7   year               344 non-null    int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 21.6+ KB


In [26]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      344.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.443643       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.275000      15.600000         190.000000  3550.000000   
50%         44.250000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

<h1>Changing Data Type</h1>

In [27]:
penguins['species'] = penguins['species'].astype('category')
penguins['island'] = penguins['island'].astype('category')

In [28]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    category
 2   bill_length_mm     344 non-null    float64 
 3   bill_depth_mm      342 non-null    float64 
 4   flipper_length_mm  342 non-null    float64 
 5   body_mass_g        342 non-null    float64 
 6   sex                344 non-null    object  
 7   year               344 non-null    int64   
dtypes: category(2), float64(4), int64(1), object(1)
memory usage: 17.2+ KB


<h1>Questions:</h1>

<h2>Compare the average bill depth betweemn Adelie and Gentoo penguins. Which is higher?</h2>

In [29]:
adelie_df = penguins[penguins['species'] == 'Adelie']['bill_depth_mm'].mean()
gentoo_df = penguins[penguins['species'] == 'Gentoo']['bill_depth_mm'].mean()
"Adelie" if adelie_df > gentoo_df else "Gentoo"

'Adelie'

<h2>Identify any species that appear on all island</h2>

<h2>Which penguin species has the longest average flipper lenght?</h2>

<h2>Average body mass by sex</h2>

In [30]:
avg_mass = penguins.groupby(['sex'])['body_mass_g'].mean()
avg_mass

sex
female    3862.272727
male      4518.220339
Name: body_mass_g, dtype: float64

<hr>
<h1>Data Aggreation</h1>

In [31]:
# body_mass sex

avg_mass = penguins.groupby(['sex'])['body_mass_g'].mean()

In [32]:
avg_mass

sex
female    3862.272727
male      4518.220339
Name: body_mass_g, dtype: float64

In [33]:
# what is the max flipper lenght by species and sex
# species, flipper lenght, sex

max_flp = penguins.groupby(['species','sex'])['flipper_length_mm'].max()
max_flp

species    sex   
Adelie     female    202.0
           male      210.0
Chinstrap  female    202.0
           male      212.0
Gentoo     female    222.0
           male      231.0
Name: flipper_length_mm, dtype: float64

<hr>
<h1>Correlation Analysis</h1>
<hr>
<h2>Determining Correlation Using Pandas Pearson R</h2>

<hr>
<h1>Questions:</h1>
<hr>

<h2>1. How many penguins belong to each species?</h2>

In [34]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

<h3>Insights:</h3>
<h3>Shows that Adelies dominate the dataset and might be easier to study.</h3>
<hr>

<h2>2. What is the mean bill length of penguins?</h2>

In [35]:
penguins['bill_length_mm'].mean()

np.float64(43.92192982456141)

<h3>Insights:</h3>
<h3>On average, penguins have bills about 4.4 cm long. This can indicate their feeding adaptation.</h3>
<hr>

<h2>3. What is the mean body mass of male vs female penguins?</h2>

In [36]:
penguins.groupby('sex')['body_mass_g'].mean()

sex
female    3862.272727
male      4518.220339
Name: body_mass_g, dtype: float64

<h3>Insights:</h3>
<h3>Males are heavier, which is typical in penguin species because of sexual dimorphism (size difference between sexes).</h3>
<hr>

<h2>4. What is the maximum body mass recorded?</h2>

In [37]:
penguins['body_mass_g'].max()

np.float64(6300.0)

<h3>Insights:</h3>
<h3>The heaviest penguin recorded weighed 6.3 kg, showing size variation in the species.</h3>
<hr>

<h2>5. How many missing values are there in each column?</h2>

In [38]:
penguins.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        2
flipper_length_mm    2
body_mass_g          2
sex                  0
year                 0
dtype: int64

<h3>Insights:</h3>
<h3>Some measurements like sex and body mass have missing data.
Shows the challenges of collecting biological data in the wild.</h3>
<hr>

<h2>6. What is the max flipper lenght by species and sex?</h2>

In [39]:
penguins.groupby(['species','sex'])['flipper_length_mm'].max()

species    sex   
Adelie     female    202.0
           male      210.0
Chinstrap  female    202.0
           male      212.0
Gentoo     female    222.0
           male      231.0
Name: flipper_length_mm, dtype: float64

<h3>Insights:</h3>
<h3>Male penguins across all species have longer flippers than females, with Gentoo males reaching the maximum length of 231 mm. This shows sexual dimorphism and highlights Gentoo as the strongest swimmers among the species.</h3>
<hr>

<h2>7. Maximum flipper length by island</h2>

In [40]:
penguins.groupby('island')['flipper_length_mm'].max()

island
Biscoe       231.0
Dream        212.0
Torgersen    210.0
Name: flipper_length_mm, dtype: float64

<h3>Insights:</h3>
<h3>Tells us the longest flipper found on each island.</h3>
<hr>

<h2>8. Mean bill dimensions (length & depth) by species and sex</h2>

In [41]:
penguins.groupby(['species','sex'])[['bill_length_mm','bill_depth_mm']].mean()

bill_length_mm  bill_depth_mm
species   sex                                  
Adelie    female       37.257534      17.621918
          male         40.273695      19.024359
Chinstrap female       46.573529      17.588235
          male         51.094118      19.252941
Gentoo    female       45.563793      14.237931
          male         49.156393      15.646154

<h3>Insights:</h3>
<h3>Males across all species have longer and deeper bills than females, with Chinstrap penguins having the longest bills and Gentoo penguins the shallowest.</h3>
<hr>

<h2>9. What is the mean bill depth of each species on each island?</h2>

penguins.groupby(['species','island'])['bill_depth_mm'].mean()

<h3>Insights:</h3>
<h3>Adelie and Chinstrap penguins have similar bill depths around 18 mm across islands, while Gentoo penguins on Biscoe stand out with a much shallower average of ~15 mm, reflecting different feeding adaptations.</h3>
<hr>

<h2>10. What is the sum of all recorded flipper lengths?</h2>

In [42]:
penguins['flipper_length_mm'].sum()

np.float64(68713.0)

<h3>Inisghts:</h3>
<h3>If we lined up all penguin flippers in the dataset, they’d measure almost 68 meters!</h3>
<hr>

<h2>11. What is the mode of penguin sex? (most common sex recorded)</h2>

In [43]:
penguins['sex'].mode()[0]

'male'

<h3>Insights:</h3>
<h3>More males than females in the dataset, which might reflect either sampling bias or real population differences.</h3>
<hr>

<h2>12. How many penguins are on each island?</h2>

In [44]:
penguins['island'].value_counts()

island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64

<h3>Insights:</h3>
<h3>Biscoe has the most penguins, followed by Dream and Torgersen. This highlights where most data collection happened.</h3>
<hr>

<h2>13. Which species-sex group has the longest average bill length?</h2>

In [45]:
penguins.groupby(['species','sex'])['bill_length_mm'].mean()

species    sex   
Adelie     female    37.257534
           male      40.273695
Chinstrap  female    46.573529
           male      51.094118
Gentoo     female    45.563793
           male      49.156393
Name: bill_length_mm, dtype: float64

<h3>Insights: </h3>
<h3>Males consistently have longer bills than females across all species. Chinstrap males have the longest average bill length.</h3>
<hr>

<h2>14. Is bill length correlated with bill depth?</h2>

In [46]:
penguins['bill_length_mm'].corr(penguins['bill_depth_mm'])

np.float64(-0.23505287035553257)

<h3>Insights:</h3>
<h3>The correlation is weak, meaning bill length and depth vary independently. Some penguins have long thin bills, while others have short deep bills.</h3>
<hr>

<h2>15. What is the correlation between flipper length and body mass?</h2>

In [47]:
penguins['flipper_length_mm'].corr(penguins['body_mass_g'])

np.float64(0.8712017673060113)

<h3>Insights:</h3>
<h3>Flipper length and body mass are strongly positively correlated. This means bigger penguins generally have longer flippers.</h3>
<hr>